File: Lab4/Lab4/Function.cpp
#include "HeaderFunction.h"
//TIntNumber
TIntNumber::TIntNumber(int number)
{
	this->number = number;
}
int TIntNumber::getSumNum()
{
	string str = to_string(number);
	int summa = 0;

	for (int i = 0; i < str.length(); i++)
	{
		summa += str[i] - '0'; 
	}

	return summa;
}
int TIntNumber::getFirstNum()
{
	string str = to_string(number);
	if(str[0] == '-') return (str[1] - '0')*(-1);
	else return str[0] - '0';
}
int TIntNumber::getLastNum()
{
	string str = to_string(number);
	return str[str.length() - 1] - '0';
}
int TIntNumber::getNumber()
{
	return number;
}


//TRealNumber
TRealNumber::TRealNumber(double number)
{
	this->number = number;
}
int TRealNumber::getSumNum()
{
	string str = to_string(number);
	int summa = 0;

	for (int i = 0; i < str.length(); i++)
	{
		if (str[i] != '.') { summa += str[i] - '0'; }
	}

	return summa;
}
int TRealNumber::getFirstNum()
{
	string str = to_string(number);
	if (str[0] == '-') return (str[1] - '0') * (-1);
	else return str[0] - '0';
}
int TRealNumber::getLastNum()
{
	string str = to_string(number);
	return str[str.length() - 1] - '0';
}
double TRealNumber::getNumber()
{
	return number;
}


//Function
vector<TIntNumber> makeIntVec(int size)
{
	srand(time(NULL));
	vector<TIntNumber> vec;
	for (int i = 0; i < size; i++)
	{
		vec.push_back((rand() % 100 - (-100) + 1) + (-100));
	}
	return vec;
}
vector<TRealNumber> makeRealVec(int size)
{
	srand(time(NULL));
	vector<TRealNumber> vec;
	for (int i = 0; i < size; i++)
	{
		vec.push_back((double)(rand())/RAND_MAX*(100 - (-100) +1) +  (-100));
	}
	return vec;
}

void printVec(vector<TIntNumber> vec)
{
	for (int i = 0; i < vec.size(); i++)
	{
		cout << vec[i].getNumber() << " ";
	}
	cout << endl;
}
void printVec(vector<TRealNumber> vec)
{
	for (int i = 0; i < vec.size(); i++)
	{
		cout << vec[i].getNumber() << " ";
	}
	cout << endl;
}

int findFirstNumSum(vector<TIntNumber> vec)
{
	int summa = 0;
	for (int i = 0; i < vec.size(); i++)
	{
		summa += vec[i].getFirstNum();
	}
	return summa;
}

int findLastNumSum(vector<TRealNumber> vec)
{
	int summa = 0;
	for (int i = 0; i < vec.size(); i++)
	{
		summa += vec[i].getLastNum();
	}
	return summa;
}

File: Lab4/Lab4_PY/Function.py
from abc import ABC, abstractmethod
import random

class TNumb:
    @abstractmethod
    def getSumNum(self):
        pass
    @abstractmethod
    def getFirstNum(self):
        pass
    @abstractmethod
    def getLastNum(self):
        pass

class TIntNumb(TNumb):
    def __init__(self, num):
        self.number = num

    def getFirstNum(self):
        string = str(self.number)
        if string[0] == '-':
            return int(string[1]) * -1
        else:
            return int(string[0])

    def getLastNum(self):
        string = str(self.number)
        return int(string[len(string) - 1])

    def getSumNum(self):
        string = str(self.number)
        summa = 0
        for i in range(len(string)):
            summa = summa + int(string[i])
        return summa
    def getNum(self):
        return self.number

class TRealNumb(TNumb):
    def __init__(self, num):
        self.number = num

    def getFirstNum(self):
        string = str(self.number)
        if string[0] == '-':
            return int(string[1]) * -1
        else:
            return int(string[0])

    def getLastNum(self):
        string = str(self.number)
        return int(string[len(string)-1])

    def getSumNum(self):
        string = str(self.number)
        summa = 0
        for i in range(len(string)):
            summa = summa + int(string[i])
        return summa

    def getNum(self):
        return self.number

def makeIntVec(size):
    arr = []
    for i in range(size):
        temp = TIntNumb(random.randint(-100,100))
        arr.append(temp)
    return arr

def makeRealVec(size):
    arr = []
    for i in range(size):
        num = random.uniform(-100,100)
        num = float('{:.1f}'.format(num))
        temp = TRealNumb(num)
        arr.append(temp)
    return arr

def printVec(arr):
    for i in range(len(arr)):
        print(arr[i].getNum(), end = ' ')
    print()

def findLastNumSum(arr):
    summ = 0
    for i in range(len(arr)):
        summ = summ + arr[i].getLastNum()
    return summ

def findFirstNumSum(arr):
    summ = 0
    for i in range(len(arr)):
        summ = summ + arr[i].getFirstNum()
    return summ
